proposals/0000-unlifted-datatypes.rst

```diff
@@ -0,0 +1,465 @@
+Fundamentally, ⊥ is an inhabitant of every data type in Haskell, so every data
+declaration introduces a *lifted* type of kind ``TYPE (BoxedRep Lifted)``.
+
+So omission of the zero tag check is only possible if ⊥ is not an inhabitant of
+the type. This is exactly the condition for a type to be *unlifted*! Since we
+are still talking about heap objects, we need unlifted, but boxed types, which
+have kind ``TYPE (BoxedRep Unlifted)``. Indeed, for types of this kind GHC will
+already omit the zero tag check today! So all that is left is to extend GHC in
+a way that
+
+::
+  
+  data IntSet :: TYPE (BoxedRep Unlifted) where
+    Branch :: IntSet -> !Int -> IntSet -> IntSet
+    Leaf   :: IntSet
+
+  member :: Int -> IntSet -> Bool
+  member !k Leaf = False
+  member k (Branch l k' r)
+    | k == k'   = True
+    | k <  k'   = member k l
+    | otherwise = member k r
+
+does the Right Thing, namely disallowing unevaluated boxes to be passed around
+by using call-by-value. This way we can have the guarantee that all unlifted
+boxes are always properly tagged.
+
+Preliminary `benchmarks <https://gitlab.haskell.org/ghc/ghc/merge_requests/2218#note_239994>`_
+suggest that the unlifted variant is 5-10% faster, just due to omission of the
+zero tag check.
+
+Proposed Change Specification
+-----------------------------
+
+Henceforth, data type declaration refers to both data type and data family instance declarations.
+
+Static semantics
+~~~~~~~~~~~~~~~~
+
+Kind signatures for regular data type declarations must have a return kind of
+``TYPE (BoxedRep Lifted)``. Activating ``-XUnliftedDatatypes`` will lift this
+restriction to allow any return kind that unifies with ``TYPE (BoxedRep _)``. A
+data type constructor explitly *may* have levity polymorphic kind
+``forall l. TYPE (BoxedRep l)``.
+
+Note that Haskell98-style data declarations can use top-level kind signatures
+to specify the return kind. If the user provides no kind signature, the default
+remains that the data type has lifted return kind (``TYPE (BoxedRep Lifted)``).
+Specifying a kind signature is the only way to declare an unlifted data type.
+
+The same applies to data family instances, in which case the data family
+application's result kind must reduce to ``TYPE (BoxedRep _)``. See
+`the section on data families in the UnliftedNewtypes proposal <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0098-unlifted-newtypes.rst>`_
+and ``Note [Implementation of UnliftedNewtypes]`` for details involving
+type-checking the parent data family.
+
+As usual for types of unlifted kind, they may not be declared at the top-level.
+The top-level restriction is not fundamental (see
+`#17521 <https://gitlab.haskell.org/ghc/ghc/issues/17521>`_), but best
+discussed in a separate proposal. There is an interesting interaction of this
+restriction with nullary data-con wrappers (cf. "Effects and Interactions"),
+which we consider to be an implementation detail.
```

Comment:
I'm not convinced here. Whether or not nullary constructors are allowed is part of this specification.